### PR TITLE
Implement Path extractPath, tangent APIs

### PR DESCRIPTION
**Pull request by ferhatb**
Related issues:
https://github.com/flutter/flutter/issues/41395
https://github.com/flutter/flutter/issues/48386

### CODE REVIEW

File: lib/web_ui/lib/src/engine/surface/path_metrics.dart [hunk 1]
```diff
@@ -524,10 +723,275 @@ class _EllipseSegmentResult {
...
+      double y1, double x2, double y2) {
+    assert(t >= 0 && t <= 1);
+    final _SkQuadCoefficients _quadEval =
+        _SkQuadCoefficients(x0, y0, x1, y1, x2, y2);
+    final ui.Offset pos = ui.Offset(_quadEval.evalX(t), _quadEval.evalY(t));
+    // Derivative of quad curve is 2(b - a + (a - 2b + c)t).
+    // If control point is at start or end point, this yields 0 for t = 0 and
+    // t = 1. In that case use the quad end points to compute tangent instead
+    // of derivative.
+    final ui.Offset tangentVector = ((t == 0 && x0 == x1 && y0 == y1) ||
```
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/surface/path_metrics.dart]: Based on comment above, shouldn't we also be checking for `t == 1` here?

File: lib/web_ui/lib/src/engine/surface/path_metrics.dart [hunk 2]
```diff
@@ -524,10 +723,275 @@ class _EllipseSegmentResult {
...
+
+  buffer[0] = abc1x;
+  buffer[1] = abc1y;
+  buffer[2] = ab2x;
+  buffer[3] = ab2y;
+  buffer[4] = abc2x;
+  buffer[5] = abc2y;
 }
+
+double _interpolate(double startValue, double endValue, double t)
```
Comment (hterkelsen) [inline lib/web_ui/lib/src/engine/surface/path_metrics.dart]: nit: I'm pretty sure this already exists somewhere (`lerp`?)

File: lib/web_ui/lib/src/engine/surface/path_metrics.dart [hunk 3]
```diff
@@ -18,84 +20,45 @@ part of engine;
...
-      return true;
-    }
-    _pathMetric = null;
-    return false;
+/// Maintains a single instance of computed segments for set of PathMetric
+/// objects exposed through iterator.
+class _SurfacePathMeasure {
+  _SurfacePathMeasure(this._path, this.forceClosed) {
+    _currentContourIndex =
+        -1; // nextContour will increment this to the zero based index.
```
[the remark this answers is not in the document]
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/surface/path_metrics.dart]: Done.

File: lib/web_ui/lib/src/engine/surface/path_metrics.dart [hunk 4]
```diff
@@ -151,27 +98,174 @@ class SurfacePathMetric implements ui.PathMetric {
...
+/// Builds segments for a single contour to measure distance, compute tangent
+/// and extract a sub path.
+class _PathContourMeasure {
+  _PathContourMeasure(this.subPath, this.forceClosed) {
+    _buildSegments();
+  }
+
+  final List<_PathSegment> _segments = [];
+  // Allocate buffer large enough for returning cubic curve chop result.
+  static final Float32List _buffer = Float32List(4 * 2);
```
[the remark this answers is not in the document]
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/surface/path_metrics.dart]: thx. added comment above.

File: lib/web_ui/lib/src/engine/surface/path_metrics.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/surface/path_metrics.dart]: thx for spotting incorrect t == 0. Corrected.

File: lib/web_ui/lib/src/engine/surface/path_metrics.dart (hunk 2 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/surface/path_metrics.dart]: Added comment. lerpDouble uses num does null checks.

**Comment by ferhatb** [comment]
Merging on infra failure.

**Pull request merged**